Tests/SwiftBSONTests/ExtendedJSONConversionTests.swift

```diff
@@ -542,4 +539,185 @@ open class ExtendedJSONConversionTestCase: BSONTestCase {
+            json: [
+                "val": [
+                    "$binary": JSON(.string(base64)),
+                    "$type": 1235
+                ]
+            ],
+            expectation: .error()
+        )
+
+        Self.jsonTest(
+            json: [
+                "val": [
+                    "$binary": JSON(.string(base64)),
+                    "$type": -123
+                ]
+            ],
+            expectation: .error()
+        )
+
+        Self.jsonTest(
+            json: [
+                "val": [
+                    "$binary": JSON(.string(base64)),
+                    "$type": true
+                ]
+            ],
+            expectation: .error()
+        )
+
+        let invalidStringSubtype: JSON = [
+            "$binary": JSON(.string(base64)),
+            "$type": "hello"
+        ]
+        Self.jsonTest(json: invalidStringSubtype, expectation: .error())
+
+        let invalidStringNumberSubtype: JSON = [
+            "$binary": JSON(.string(base64)),
+            "$type": "12345"
+        ]
+        Self.jsonTest(json: invalidStringNumberSubtype, expectation: .error())
+    }
+
+    func testLegacyExtendedJSONDate() throws {
+        Self.jsonTest(json: ["val": ["$date": 0]], expectation: .success(Date(timeIntervalSince1970: 0)))
+        Self.jsonTest(
+            json: ["val": ["$date": 1_356_351_330_500]],
+            expectation: .success(Date(msSinceEpoch: 1_356_351_330_500))
+        )
+        Self.jsonTest(
+            json: ["val": ["$date": -62_135_593_139_000]],
+            expectation: .success(Date(msSinceEpoch: -62_135_593_139_000))
+        )
+        Self.jsonTest(
+            json: ["val": ["$date": JSON(.number(String(Int64.max)))]],
+            expectation: .success(Date(msSinceEpoch: Int64.max))
+        )
+        Self.jsonTest(
+            json: ["val": ["$date": JSON(.number(String(Int64.min)))]],
+            expectation: .success(Date(msSinceEpoch: Int64.min))
+        )
+        // Int64.max + 1
+        Self.jsonTest(
+            json: ["val": ["$date": JSON(.number("9223372036854775808"))]],
+            expectation: .error()
+        )
+        // Int64.min - 1
+        Self.jsonTest(
+            json: ["val": ["$date": JSON(.number("-9223372036854775809"))]],
+            expectation: .error()
+        )
+        Self.jsonTest(
+            json: ["val": ["$date": JSON(.number("10000000000000000000"))]],
+            expectation: .error()
+        )
+        Self.jsonTest(
+            json: ["val": ["$date": JSON(.number("-10000000000000000000"))]],
+            expectation: .error()
+        )
+    }
+
+    func testLegacyExtendedJSONRegex() throws {
+        let regex = BSONRegularExpression(pattern: "abc", options: "ix")
+
+        Self.jsonTest(
+            json: ["val": ["$regex": "abc", "$options": "ix"]],
+            expectation: .success(regex)
+        )
+
+        // don't invalidate a "$regex" query operator stored in JSON
+        Self.jsonTest(
```

Comment:
the first of these 3 were taken from libbson too, except libbson would return an error. That seemed problematic to me, since these could easily be valid `$regex` query operators, just stored in JSON. I figured no longer throwing an error in these cases would not really constitute a breaking change, and that the legacy parsing should defer to the non-legacy path if it ever is ambiguous.